### PR TITLE
[V2] Deferred Props

### LINKED ISSUE #129: V2: Implementing Deferred props
V2 is coming to Inertia, so let's get ready.

## Context

The concept of Deferred props is:  

> Props that are not evaluated on initial page load but are requested immediately after a page is loaded, automatically.

On the backend, they work similarly to LazyProps, but the key differences lie in how they're handled on the frontend.

## Implementation

Probably we can just create an alias to `InertiaRails.lazy` and create this new feature for grouping props

### Group

From _advanced-inertia.com_
> Inertia also lets you group deferred props. This enables you to load certain deferred groups independently from others, so that quicker queries will not have to wait until the slower ones are completed.

We have to implement a group feature too

Source:
- https://advanced-inertia.com/blog/inertia-v2
- Laravel Implementation: https://github.com/inertiajs/inertia-laravel/compare/1.x...async-requests#diff-e9c6c2799322c38d5afd5376eea239f415bc158b66fba4a7591ddaf8bb93bcc5

### PR DESCRIPTION
**Pull request by PedroAugustoRamalhoDuarte**
Initial PR to implement Deferred Props in rails

## TODO
- [x] Create an interface to handle new IgnoreInFirstLoadProps methods
- [x] Create `InertiaRails.defer` method
- [x] Group feature for `InertiaRails.defer`
- [x] Share the Deffered Props inside deferredProps inertia response

closes #129 

inspired by:
- https://github.com/inertiajs/inertia-laravel/compare/1.x...async-requests

### CODE REVIEW

**Comment by buhrmi** [comment]
i will test it as soon as the 2.0 beta is out

File: lib/inertia_rails/ignore_first_load_prop.rb [hunk 1]
```diff
@@ -1,5 +1,5 @@
 module InertiaRails
-  class Lazy
+  class IgnoreFirstLoadProp
```
Comment (skryukov) [inline lib/inertia_rails/ignore_first_load_prop.rb:2]: I don't think we should rename `Lazy`, let's create a base class `IgnoreFirstLoadProp` instead to mimic what Laravel's plugin does with the interface?

File: lib/inertia_rails/renderer.rb [hunk 2]
```diff
@@ -77,6 +82,15 @@ def deep_transform_values(hash, proc)
       hash.transform_values {|value| deep_transform_values(value, proc)}
     end
 
+    def deferred_props_keys
+      return if rendering_partial_component?
+
+      @props.select { |_, prop| prop.is_a?(DeferProp) }
```
Comment (skryukov) [inline lib/inertia_rails/renderer.rb:88]: while I really love the functional approach, personally, I would optimize this with smth like:
```ruby
  @props.each_with_object({}) do |(key, prop), result|
    (result[prop.group] ||= []) << key.to_s if prop.is_a?(DeferProp)
  end
```

**Review by skryukov — COMMENTED** [review]
Thanks @PedroAugustoRamalhoDuarte 🙌 I left a couple of comments, otherwise it looks really good!

File: lib/inertia_rails/ignore_first_load_prop.rb (hunk 1 above)
Comment (PedroAugustoRamalhoDuarte) [reply lib/inertia_rails/ignore_first_load_prop.rb:2]: I like the idea too, IgnoreFirstLoad too big, and i think optional is the new nomeclature

File: lib/inertia_rails/renderer.rb (hunk 2 above)
Comment (PedroAugustoRamalhoDuarte) [reply lib/inertia_rails/renderer.rb:88]: Yes much better

**Comment by PedroAugustoRamalhoDuarte** [comment]
@skryukov i don't know with this PR make sense alone right now that is merged into v2

**Comment by PedroAugustoRamalhoDuarte** [comment]
Closing in favor of https://github.com/inertiajs/inertia-rails/pull/132

**Pull request closed without merging**